Update  chapter 15 (#246)
hjwp/Book-TDD-Web-Dev-Python: chapter_15_advanced_forms.asciidoc

----
class Item(models.Model):
    text = models.TextField(default='')
    list = models.ForeignKey(List, default=None, on_delete=models.CASCADE)

    class Meta:
        unique_together = ('list', 'text')
----
====

You might want to take a quick peek at the
https://docs.djangoproject.com/en/4.2/ref/models/options/[Django docs on model
`Meta` attributes] at this point.

Great! Our unit tests are passing.
----
OK
----

> // CSANAD: fully passing suite? I think just the unit tests

Reviewer: 👍

// TODO: in theory we should do a migration now.

[[rewrite-model-test]]
==== Rewriting the Old Model Test

That long-winded model test did serendipitously help us find unexpected
bugs, but now it's time to rewrite it. I wrote it in a very verbose style to
introduce the Django ORM, but in fact, we can get the same coverage from a
couple of much shorter tests.

> Delete `test_saving_and_retrieving_items` and replace with this:

Reviewer: ```suggestion
Delete `test_saving_and_retrieving_items` and replace it with this:
```

$ pass:quotes[*python manage.py makemigrations*]
Migrations for 'lists':
  src/lists/migrations/0005_alter_item_options_alter_item_unique_together.py
    - Alter unique_together for item (1 constraint(s))
----

Now let's run the migration:

[subs="specialcharacters,macros"]
----
$ pass:quotes[*python manage.py migrate*]
----

When you run the migration, you may encounter the following error:
[subs="specialcharacters,macros"]
----
$ pass:quotes[*python manage.py migrate*]
Operations to perform:
  Apply all migrations: auth, contenttypes, lists, sessions
Running migrations:
  Applying lists.0005_alter_item_unique_together...Traceback (most recent call last):
[...]
sqlite3.IntegrityError: UNIQUE constraint failed: lists_item.list_id, lists_item.text

[...]
django.db.utils.IntegrityError: UNIQUE constraint failed: lists_item.list_id, lists_item.text
----

The problem is that we have at least one database record which used to be valid
but after introducing our new constraint, the `unique_together`, it's no longer
compatible.

> Just delete `src/db.sqlite3` and run the migration again. We can do this,
> because it's only a dev environment, and the data saved in the database isn't
> important.

Reviewer: ```suggestion
To fix this problem, we can just delete `src/db.sqlite3` and run the migration again.
We can do this because the database on our laptop is only used for dev, so the data in it is not important.
```

However, if we already have records we want to keep (which would be the case in
a production environment, where we need to keep our users' data) we would need
to do a bit more work and come up with a migration strategy for dealing with
invalid--duplicate in this case--records.
// CSANAD: I wanted to mention some articles from credible sources explaining
//         migration strategies, modifying migration scripts, etc
//         but all I found were either outdated, or clearly chatgpt-generated

> // articles.

Reviewer: hmmm i think we do mention prod migrations at some point... let me look it up.

Reviewer: yes it's mentioned in chapter 17, when we try and deploy the latest code to prod, and there's an appendix on data migrations.  http://www.obeythetestinggoat.com/book/chapter_deploying_validation.html#_what_to_do_if_you_see_a_database_error

Reviewer: ```suggestion
In <chapter_deploying_validation>>, we'll deploy our new code to production,
and discuss what to do if we run into migrations and data integrity issues at that point.
```

-         {{ form.text }}
+     <form
+         class="was-validated"  <1>
+         method="POST"
+         action="{% block form_action %}{% endblock %}"
+     >
          {% csrf_token %}
+         <input
+             id="id_text"
+             name="text"
+             class="form-control form-control-lg {% if form.errors %}is-invalid{% endif %}"  <2>
+             placeholder="Enter a to-do item"
+             value="{{ form.text.value }}"
+             aria-describedby="id_text_feedback"
+             required
+         />
          {% if form.errors %}
-             <div class="invalid-feedback">{{ form.errors.text }}</div>
+             <div class="invalid-feedback">{{ form.errors.text.0 }}</div>  <3>
          {% endif %}
      </form>
  </div>
----
====

<1> We give `.was-validated` to the `<form>`
<2> We provide an `<input>` and the most important custom setting will be its
    `class`. As you can see, we can use conditionals even for providing
    additional `class` -es.
<3> If you try to just simply display `form.errors.text` you'll see it ends up
    as a list item in an unordered list. To avoid that, let's just
    select `form.errors.text.0`.

> 

Reviewer: ```suggestion

NOTE:  Another fliip-flop! We spent most of the last chapter switching from handcrafted html
  to having our form autogenerated by django, and now we're switching back.
  It's a little frustrating, and I could have gone back and changed the book's text to avoid the back+forth,
  but I prefer to show software development as it really is.
  We often try things out and end up changing our minds.
  Particularly with frameworks like Django,
  you can find yourself taking advantage of auto-generated shortcuts for as long as they work,
  but at some points you meet the limits of what the framework designers have anticipated,
  and it's time to go back to doing the work yourself.
  It doesn't mean you should always reinvent the wheel!
```
